ui/StatusQ/src/singleModelItemData.cpp

```diff
@@ -0,0 +1,254 @@
+            setIndex(index);
+            return;
+        }
+        
+        // Check if the index is still valid
+        auto index = findIndexInRange(m_index.row(), m_index.row() + 1);
+        if (index != m_index)
+        {
+            setIndex(index);
+            return;
+        }
+
+        if (topLeft.row() <= m_index.row() && m_index.row() <= bottomRight.row())
+        {
+            updateItem(roles);
+        }
+    });
+    // connect(m_sourceModel, &QAbstractItemModel::layoutChanged, this, [this]() {
+    //     if (!m_index.isValid())
+    //     {
+    //         resetIndex();
+    //         return;
+    //     }
+        
+    //     auto index = findIndexInRange(m_index.row(), m_index.row() + 1);
+    //     setIndex(index);
+    // });
+
+    emit sourceModelChanged();
+}
+
+void SingleModelItemData::setFilter(SingleItemFilter* filter)
+{
+    if (m_filter == filter)
+        return;
+
+    if (m_filter)
+        disconnect(m_filter, nullptr, this, nullptr);
+
+    m_filter = filter;
+
+    if (m_filter)
+    {
+        resetIndex();
+        connect(m_filter, &SingleItemFilter::invalidate, this, [this]() {
+            resetIndex();
+        });
+    }
+
+    emit filterChanged();
+}
+
+void SingleModelItemData::setIndex(const QModelIndex& index)
+{
+    if (m_index == index)
+        return;
+
+    m_index = index;
+
+    tryItemResetOrUpdate();
+    setAvailable(m_index.isValid());
+    emit rowChanged();
+}
+
+void SingleModelItemData::setAvailable(bool available)
+{
+    if (available == m_available)
+        return;
+
+    m_available = available;
+    emit availableChanged();
+}
+
+void SingleModelItemData::setRoles(const QStringList& roles)
+{
+    if (m_roles == roles)
+        return;
+
+    m_roles = roles;
+    emit rolesChanged();
+}
+
+QModelIndex SingleModelItemData::findIndexInRange(int start, int end) const
+{
+    if (!m_sourceModel || !m_filter)
+        return {};
+
+    return m_filter->findIndexInRange(m_sourceModel, start, end);
+}
+
+void SingleModelItemData::resetIndex()
+{
+    if (!m_sourceModel || !m_filter)
+        return;
+
+    auto index = findIndexInRange(0, m_sourceModel->rowCount());
+    setIndex(index);
+}
+
+void SingleModelItemData::tryItemResetOrUpdate()
+{
+    if (!m_sourceModel)
+        return;
+
+    if (!m_index.isValid() || !itemHasCorrectRoles())
+    {
+        resetItem();
+        return;
+    }
+
+    updateItem();
+}
+
+void SingleModelItemData::resetItem()
+{
+    m_item = new QQmlPropertyMap(this);
+
+    updateItem();
+    setRoles(m_item->keys());
+
+    emit itemChanged();
+}
+
+void SingleModelItemData::updateItem(QList<int> roles)
+{
+    if (!m_index.isValid())
+        return;
+
+    if (roles.isEmpty())
+    {
+        auto roleNames = m_sourceModel->roleNames();
+        roles = roleNames.keys();
+    }
+
+    for (auto role : roles)
+    {
+        auto roleName = m_sourceModel->roleNames().value(role);
```

Comment:
It would be good to keep source model's role name cached. Frequent asking about role may be a bottleneck, especially for models with poorly implemented `roleNames`.